compiler/rustc_resolve/src/late.rs

```
@@ -3739,7 +3741,62 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
                 )));
             }
 
-            // Make sure `A::B` in `<T as A::B>::C` is a trait item.
+            // Make sure that `A` in `<T as A>::B::C` is a trait.
+            //
+            // This is best-effort, for diagnostics, as the subsequent
+            // call to `smart_resolve_path_fragment` will fail even if
+            // we don't emit an error here.
+            let trait_path = &path[..qself.position];
+            match self.resolve_path(trait_path, Some(TypeNS), None) {
+                PathResult::NonModule(res) => {
+                    match res.full_res() {
+                        Some(Res::Def(DefKind::Trait, _)) => unreachable!(),
+                        Some(Res::Def(def_kind, def_id)) => {
+                            return Err(Spanned {
```

I guess we could report the error eagerly rather than bubbling it up?